examples/centralized_telescope.rs

```diff
@@ -0,0 +1,146 @@
+//! Example
+
+// REMOVE!!!!!!!!!!!!!!!!
+#![allow(dead_code)]
+
+mod threshold_signature;
+
+use alba::centralized_telescope::params::Params;
+use alba::centralized_telescope::proof::Proof;
+use alba::centralized_telescope::CentralizedTelescope;
+use rand_chacha::ChaCha20Rng;
+use rand_core::{RngCore, SeedableRng};
+
+use crate::threshold_signature::registration::{ClosedRegistration, RegisteredKeys};
+use crate::threshold_signature::signature::IndividualSignature;
+use crate::threshold_signature::signer::{Candidate, Signer};
+use threshold_signature::aggregate::AggregateSignature;
+
+const DATA_LENGTH: usize = 32;
+pub(crate) type Element = [u8; DATA_LENGTH];
+
+/// Alba proof with aggregate signature
+#[derive(Debug, Clone)]
+pub struct AlbaThresholdProof {
+    /// Aggregate signature
+    pub(crate) aggregate: AggregateSignature,
+    /// Centralized telescope proof
+    pub proof: Proof,
+}
+
+impl AlbaThresholdProof {
+    /// Create an Alba proof
+    /// - Try to aggregate given list of aggregate signatures
+    /// - If aggregation succeeds, create the Alba proof
+    /// - If Alba proof is generated, return the aggregate signature and the Alba proof
+    pub(crate) fn prove<const N: usize>(
+        params: &Params,
+        signatures: &[IndividualSignature],
+        set_size: u64,
+        closed_registration: &ClosedRegistration,
+        msg: &[u8],
+    ) -> Option<Self> {
+        let try_aggregate =
+            AggregateSignature::aggregate::<N>(signatures, closed_registration, msg, set_size);
+        if let Some(aggregate) = try_aggregate {
+            let prover_set: Vec<Element> = aggregate.create_prover_set::<N>();
+            let alba = CentralizedTelescope::create(params);
+            let try_proof = alba.prove(&prover_set);
+            if let Some(proof) = try_proof {
+                Some(Self { aggregate, proof })
+            } else {
+                println!("Proof generation failed.");
+                None
+            }
+        } else {
+            println!("Aggregation failed.");
+            None
+        }
+    }
+
+    /// Verify given Alba proof
+    /// Create the commitment by hashing the checksum of the closed registration and the message
+    /// If the computed commitment is different than the commitment of given aggregate signature, abort
+    /// Verify each individual signature of the aggregate signature. Abort if the aggregate includes an invalid signature
+    /// Return true if Alba proof is verified and all checks passed.
+    pub(crate) fn verify<const N: usize>(
+        &self,
+        params: &Params,
+        closed_registration: &ClosedRegistration,
+        msg: &[u8],
+    ) -> bool {
+        let commitment: [u8; N] =
+            ClosedRegistration::get_commitment(&closed_registration.check_sum, msg);
+
+        if commitment != self.aggregate.commitment.as_slice() {
+            return false;
+        }
+
+        for sig in self.aggregate.valid_signatures.clone() {
```

Comment:
since you are not modifying the `valid_signatures` during iteration, it is more efficient and more idiomatic to use `iter()` instead of `clone()`, that is 
```
if !self.aggregate.valid_signatures
    .iter()
    .all(|sig| sig.verify::<N>(&commitment, closed_registration))
{
    return false;
}
```

Reply:
Since all these signatures are BLS signatures you could batch verify them!

You could implement this in `aggregate.rs` using blst's function 
```
pub fn aggregate_verify(
                &self,
                sig_groupcheck: bool,
                msgs: &[&[u8]],
                dst: &[u8],
                pks: &[&PublicKey],
                pks_validate: bool,
            ) -> BLST_ERROR 
```